warehouse/macaroons/services.py

```diff
@@ -67,30 +56,22 @@ def find_macaroon(self, macaroon_id):
         return dm
 
     def _deserialize_raw_macaroon(self, raw_macaroon):
-        raw_macaroon = self._extract_raw_macaroon(raw_macaroon)
-
-        if raw_macaroon is None:
-            raise InvalidMacaroon("malformed or nonexistent macaroon")
-
         try:
-            return pymacaroons.Macaroon.deserialize(raw_macaroon)
-        except (
-            MacaroonDeserializationException,
-            Exception,  # https://github.com/ecordell/pymacaroons/issues/50
-        ):
-            raise InvalidMacaroon("malformed macaroon")
+            return pypitoken.Token.load(raw_macaroon)
+        except pypitoken.LoaderError as exc:
+            raise InvalidMacaroon(str(exc))
```

Comment:
It's one of the promises of the pypitoken API that the exception message is displayable to the user (same when calling `check()`)